### DISCUSSION TITLE
telnet request.all returns insertion order, not play order

### DISCUSSION THREAD

**Discussion by danbo** [savonet/liquidsoap#4093] [Q&A] [answered]
Hello,

As part of debugging #4071 and #3134, I noticed that the telnet request.all returns requests in insertion order and not anticipated play order.

ie 
- start with fallback([q1, q2]) where q1 empty, q2 contains r1
- push q1r1
- push q1r2
- call request.all
- it returns the push order: [0(q2r1), 1(q1s1), 2(q1s2)]
- but actual play order per fallback prioritization is: [1(q1r1), 2(q1r2), 0(q2r1)]

Is this because in order to show play order it would have to pre-calculate / calculate upon calling request.all? Is this also why it is not exposed programmatically?

**Comment by vitoyucepi** [comment]
Hi @danbo,

I think the telnet function `request.all` returns all requests in the system in order of creation. You should define your own function to get all requests, for example playlists has a `queue()` method.

**Reply by danbo** [reply]
I should have clarified that my q1 and q2 are request.dynamic.

In any case, I can get the queues individually, but I can't get the holistic order because I don't control the play order, liquidsoap does. Ideally, I should be able to query programmatically and report back to a UI what will actually be played.

**Comment by toots** [comment]
Yeah, `request.all` does not make assumptions about play order. It will also return _all_ requests, meaning it'll merge requests from multiple sources if you have several sources using requests.

If you need play order, you should query the corresponding source for its own internal queue of requests.

**Reply by toots** [reply, marked as answer]
Sorry I forgot to add: the play order itself is complicated. For instance, if you have a switch, it'll have to follow the logic you put in its predicates and etc. There's no good way in general to look at a script and tell what is the next request to be played.

**Reply by danbo** [reply]
Thank you for the info.
